hpx/parallel/datapar/detail/iterator_helpers.hpp

```diff
@@ -0,0 +1,312 @@
+                return hpx::util::invoke(f, &tmp);
+            }
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V*)>::type
+            callv(F && f, Iter& it)
+            {
+                store_on_exit<Iter, V> tmp(it);
+                it += V::Size;
+                return hpx::util::invoke(f, &tmp);
+            }
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Iter1, typename Iter2>
+        struct datapar_loop_step2
+        {
+            typedef typename std::iterator_traits<Iter1>::value_type value1_type;
+            typedef typename std::iterator_traits<Iter2>::value_type value2_type;
+
+            typedef Vc::Scalar::Vector<value1_type> V11;
+            typedef Vc::Scalar::Vector<value2_type> V12;
+
+            typedef Vc::Vector<value1_type> V1;
+            typedef Vc::Vector<value2_type> V2;
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V11*, V12*)>::type
+            call1(F && f, Iter1& it1, Iter2& it2)
+            {
+                V11 tmp1(std::addressof(*it1), Vc::Aligned);
+                V12 tmp2(std::addressof(*it2), Vc::Aligned);
+                std::advance(it1, V11::Size);
+                std::advance(it2, V12::Size);
+                return hpx::util::invoke(f, &tmp1, &tmp2);
+            }
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V1*, V2*)>::type
+            callv(F && f, Iter1& it1, Iter2& it2)
+            {
+                if (data_alignment(it1) || data_alignment(it2))
+                {
+                    V1 tmp1(std::addressof(*it1), Vc::Unaligned);
+                    V2 tmp2(std::addressof(*it2), Vc::Unaligned);
+                    std::advance(it1, V1::Size);
+                    std::advance(it2, V2::Size);
+                    return hpx::util::invoke(f, &tmp1, &tmp2);
+                }
+
+                V1 tmp1(std::addressof(*it1), Vc::Aligned);
+                V2 tmp2(std::addressof(*it2), Vc::Aligned);
+                std::advance(it1, V1::Size);
+                std::advance(it2, V2::Size);
+                return hpx::util::invoke(f, &tmp1, &tmp2);
+            }
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        struct datapar_transform_loop_step
+        {
+            template <typename F, typename InIter, typename OutIter>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static void call1(F && f, InIter& it, OutIter& dest)
+            {
+                typedef typename std::iterator_traits<InIter>::value_type
+                    value_type;
+
+                typedef Vc::Scalar::Vector<value_type> V1;
+
+                V1 tmp(std::addressof(*it), Vc::Aligned);
```

Comment:
This is a flaw then. There are several locations in the loop, where loop_optimized returns false and the policy is switch to `seq` (For example [here](https://github.com/STEllAR-GROUP/hpx/blob/10b7b35a459c00f73cbda1ec8d70bda398311cae/hpx/parallel/algorithms/inner_product.hpp#L95)) The passed function object has to work with built in types in either case.
This can be worked around with my proposal for `datapar_traits` to provide the necessary overloads.